src/rail_gui.cpp

```diff
@@ -2011,7 +2011,7 @@ struct BuildRailWaypointWindow : PickerWindowBase {
 		this->FinishInitNested(TRANSPORT_RAIL);
 
 		matrix->SetCount(_waypoint_count);
-		matrix->SetClicked(_cur_waypoint_type);
+		matrix->SetClicked(_cur_waypoint_type < _waypoint_count ? _cur_waypoint_type : 0);
```

Comment:
But `_cur_waypoint_type` will still contain an invalid value, and it's used again in `OnClick()`.